From 82dc1eddee80cc65901fe46512442b677e2dd543 Mon Sep 17 00:00:00 2001
From: Christophe Chapuis <chris.chapuis@gmail.com>
Date: Tue, 20 Oct 2015 20:33:24 +0200
Subject: [PATCH] Store the additional window features related to LuneOS in the
 WebContentsView object

 - additional_features (from window.open parameter)
 - initial target URL

Signed-off-by: Christophe Chapuis <chris.chapuis@gmail.com>
---
 .../content/browser/web_contents/web_contents_impl.cc     | 8 +++++++-
 chromium/content/browser/web_contents/web_contents_view.h | 8 ++++++++
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/chromium/content/browser/web_contents/web_contents_impl.cc b/chromium/content/browser/web_contents/web_contents_impl.cc
index b96ef72a05..6b9461d64e 100644
--- a/chromium/content/browser/web_contents/web_contents_impl.cc
+++ b/chromium/content/browser/web_contents/web_contents_impl.cc
@@ -2354,6 +2354,11 @@ void WebContentsImpl::CreateNewWindow(
     if (!is_guest) {
       WebContentsView* new_view = new_contents->view_.get();
 
+      // set the additional features required by the LuneOS app
+      // (ideally this information should be propagated using the IPC messaging)
+      new_view->setWindowAdditionalFeatures(params.additional_features);
+      new_view->setInitialTargetURL(params.target_url);
+
       // TODO(brettw): It seems bogus that we have to call this function on the
       // newly created object and give it one of its own member variables.
       new_view->CreateViewForWidget(
@@ -2491,8 +2496,9 @@ void WebContentsImpl::ShowCreatedWindow(int process_id,
     if (delegate) {
       base::WeakPtr<WebContentsImpl> weak_popup =
           popup->weak_factory_.GetWeakPtr();
+      std::vector<base::string16> additional_features = contents->view_.get()->getWindowAdditionalFeatures();
       delegate->AddNewContents(this, popup, disposition, initial_rect,
-                               user_gesture, nullptr);
+                               user_gesture, nullptr, additional_features);
       if (!weak_popup)
         return;  // The delegate deleted |popup| during AddNewContents().
     }
diff --git a/chromium/content/browser/web_contents/web_contents_view.h b/chromium/content/browser/web_contents/web_contents_view.h
index a38a936af6..dbdbb483b7 100644
--- a/chromium/content/browser/web_contents/web_contents_view.h
+++ b/chromium/content/browser/web_contents/web_contents_view.h
@@ -13,6 +13,7 @@
 #include "ui/gfx/geometry/rect.h"
 #include "ui/gfx/geometry/size.h"
 #include "ui/gfx/native_widget_types.h"
+#include "url/gurl.h"
 
 namespace content {
 class RenderViewHost;
@@ -136,6 +137,13 @@ class WebContentsView {
   virtual bool IsEventTracking() const = 0;
   virtual void CloseTabAfterEventTracking() = 0;
 #endif
+
+  // Add accessors to set/get the window additional features wanted by the LuneOS apps
+  // The actual implementation is done in web_contents_view_qt.h
+  virtual void setWindowAdditionalFeatures(const std::vector<base::string16> &additional_features) {}
+  virtual std::vector<base::string16> getWindowAdditionalFeatures() {}
+  virtual void setInitialTargetURL(const GURL &initialURL) {}
+  virtual GURL getInitialTargetURL() {}
 };
 
 }  // namespace content
